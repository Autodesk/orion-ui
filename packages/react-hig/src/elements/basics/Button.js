/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as HIG from 'hig.web';

import HIGElement from '../HIGElement';
import * as PropTypes from 'prop-types';
import createComponent from '../../adapters/createComponent';


class Button extends HIGElement {
  constructor(initialProps) {
    super(HIG.Button, initialProps);
  }

  componentDidMount(){
    this.commitUpdate(this.props);
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    this.processUpdateProps(updatePayload)
      .mapToHIGFunctions({
        title: 'setTitle',
        link: 'setLink',
        icon: 'setIcon',
        type: 'setType',
        size: 'setSize'
      })
      .mapToHIGEventListeners(['onClick', 'onHover', 'onFocus', 'onBlur'])
      .handle('disabled', value => {
        value ? this.hig.disable() : this.hig.enable();
      })

  }
}



const ButtonComponent = createComponent(Button);

Button.propTypes = {
  disabled: PropTypes.bool,
  link: PropTypes.string,
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onHover: PropTypes.func,
  title: PropTypes.string,
  size: PropTypes.string,
  type: PropTypes.string, 
  icon: PropTypes.string
}

ButtonComponent.__docgenInfo = {
  props: {
    title: {
      description: 'sets the title of a button'
    },

    link: {
      description: 'sets the link of a button'
    },

    onClick: {
      description: 'triggers when you click the button'
    },

    onHover: {
      description: 'triggers when you hover over the button'
    }
  }
};



export default ButtonComponent;
