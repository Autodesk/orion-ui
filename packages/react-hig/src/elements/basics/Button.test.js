/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import Button from './Button';

describe('<Button>', () => {
  /**
     * Creates a hig.web button and returns the instance and it's container
     *
     * @param {object} defaults hig.web defaults for the button
     */
  function createHigButton(defaults = {}) {
    const higContainer = document.createElement('div');

    // use spread here to clone defaults since HIG.Button mutates this object
    const higButton = new HIG.Button({ ...defaults });

    higButton.mount(higContainer);

    return { higButton, higContainer };
  }

  it('renders the standard button', () => {
    const defaults = { title: 'regular button', link: 'http://example.com' };

    const { higButton, higContainer } = createHigButton(defaults);
    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the small button', () => {
    const defaults = {
      size: 'small',
      type: 'primary',
      title: 'small button',
      link: 'http://example.com'
    };

    const { higButton, higContainer } = createHigButton(defaults);

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the large button', () => {
    const defaults = {
      size: 'large',
      type: 'primary',
      title: 'Large button',
      link: 'http://example.com'
    };

    const { higButton, higContainer } = createHigButton(defaults);

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the primary button', () => {
    const defaults = {
      size: 'standard',
      type: 'primary',
      title: 'Primary button',
      link: 'http://example.com'
    };

    const { higButton, higContainer } = createHigButton(defaults);

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the secondary button', () => {
    const defaults = {
      size: 'standard',
      type: 'secondary',
      title: 'Secondary button',
      link: 'http://example.com'
    };

    const { higButton, higContainer } = createHigButton(defaults);

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the flat button', () => {
    const defaults = {
      type: 'flat',
      size: 'standard',
      title: 'Flat button',
      link: 'http://example.com'
    };

    const { higButton, higContainer } = createHigButton(defaults);

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the button with an icon', () => {
    const defaults = { icon: 'gear' };

    const { higButton, higContainer } = createHigButton(defaults);

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('updates the title the same as setTitle', () => {
    const defaults = { title: 'some title' };
    const { higButton, higContainer } = createHigButton(defaults);

    const newTitle = 'correct title';

    // update via hig API
    higButton.setTitle(newTitle);

    const container = document.createElement('div');

    // update via React API
    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    wrapper.setProps({ title: newTitle });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('removes the title if it is no longer specified', () => {
    const { higButton, higContainer } = createHigButton();

    const reactContainer = document.createElement('div');

    const wrapper = mount(<Button />, {
      attachTo: reactContainer
    });

    higButton.setTitle(undefined);

    // update to remove the title prop
    wrapper.setProps({ title: undefined });

    expect(reactContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(reactContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('updates the link the same as setLink', () => {
    const { higButton, higContainer } = createHigButton();
    const newLink = 'http://example.com/2';

    higButton.setLink(newLink);

    const reactContainer = document.createElement('div');

    const wrapper = mount(<Button />, { attachTo: reactContainer });

    wrapper.setProps({ link: newLink });

    expect(reactContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(reactContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('updates the size using setSize', () => {
    const defaults = { title: 'some title', size: 'small', type: 'primary' };
    const { higButton, higContainer } = createHigButton(defaults);
    const newSize = 'large';

    higButton.setSize(newSize);

    const reactContainer = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, {
      attachTo: reactContainer
    });

    wrapper.setProps({ size: newSize });

    expect(reactContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(reactContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('updates the icon using setIcon', () => {
    const { higButton, higContainer } = createHigButton();
    const newIcon = 'hamburger';

    higButton.setIcon(newIcon);

    const reactContainer = document.createElement('div');

    const wrapper = mount(<Button />, { attachTo: reactContainer });

    wrapper.setProps({ icon: newIcon });

    expect(reactContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(reactContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders the disabled button', () => {
    const defaults = {
      size: 'standard',
      title: 'regular button',
      link: 'http://example.com',
      disabled: 'true'
    };

    const { higButton, higContainer } = createHigButton(defaults);
    higButton.disable();

    const container = document.createElement('div');

    const wrapper = mount(<Button {...defaults} />, { attachTo: container });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  ['onClick', 'onHover', 'onFocus', 'onBlur'].forEach(eventName => {
    it(`sets up ${eventName} initially`, () => {
      const eventSpy = jest.fn();

      const reactContainer = document.createElement('div');

      const wrapper = mount(<Button {...{ [eventName]: eventSpy }} />, {
        attachTo: reactContainer
      });

      // click on the rendered button
      const instance = wrapper.instance().instance;

      // This is the same fn we pass to hig.web
      instance.events[eventName]();

      // expect onClickSpy to be called
      expect(eventSpy).toBeCalled();
    });

    it(`updates ${eventName} when it updates`, () => {
      const eventSpy1 = jest.fn();
      const eventSpy2 = jest.fn();

      const reactContainer = document.createElement('div');

      const wrapper = mount(<Button {...{ [eventName]: eventSpy1 }} />, {
        attachTo: reactContainer
      });

      wrapper.setProps({ [eventName]: eventSpy2 });

      // click on the rendered button
      const instance = wrapper.instance().instance;

      // This is the same fn we pass to hig.web
      instance.events[eventName]();

      // expect onClickSpy to be called
      expect(eventSpy2).toBeCalled();
    });

    it(`updates ${eventName} when it wasn't specified to begin with`, () => {
      const eventSpy1 = jest.fn();

      const reactContainer = document.createElement('div');

      const wrapper = mount(<Button />, {
        attachTo: reactContainer
      });

      wrapper.setProps({ [eventName]: eventSpy1 });

      const instance = wrapper.instance().instance;

      // This is the same fn we pass to hig.web
      instance.events[eventName]();

      // expect onClickSpy to be called
      expect(eventSpy1).toBeCalled();
    });

    it(`removes ${eventName} if it is no longer specified`, () => {
      const eventSpy1 = jest.fn();

      const reactContainer = document.createElement('div');

      const wrapper = mount(<Button {...{ [eventName]: eventSpy1 }} />, {
        attachTo: reactContainer
      });

      wrapper.setProps({ [eventName]: undefined });

      const instance = wrapper.instance().instance;

      expect(instance.events[eventName]).toBeUndefined();
    });
  });
});
