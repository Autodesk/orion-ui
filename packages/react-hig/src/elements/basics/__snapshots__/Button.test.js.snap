// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> removes the title if it is no longer specified 1`] = `"<a href=\\"#\\" class=\\"hig__button\\" title=\\"some title\\"></a>"`;

exports[`<Button> renders the standard button 1`] = `"<a href=\\"http://example.com\\" class=\\"hig__button\\" title=\\"regular button\\">regular button</a>"`;

exports[`<Button> updates the link the same as setLink 1`] = `"<a href=\\"http://example.com/2\\" class=\\"hig__button\\" title=\\"link\\">link</a>"`;

exports[`<Button> updates the title the same as setTitle 1`] = `"<a href=\\"#\\" class=\\"hig__button\\" title=\\"link\\">correct title</a>"`;
